### ISSUE TITLE
Add webassembly-micro-runtime support

### ISSUE THREAD

**Issue by squillace** [containerd/runwasi#337]
Currently, there is shim support for wasmtime, wasmedge, and wasmer thanks to contributors from around the world. I'd like to propose that we add support for the https://github.com/bytecodealliance/wasm-micro-runtime to ensure wide coverage for runtimes in this space. Each one of these runtime specializes in something important that people should have access to anywhere they want to use K8s and WebAssembly.

**Comment by 0xE282B0** [comment]
Ok, we can do that. I already started with an executor some time ago: https://github.com/0xE282B0/runwasi-youki/tree/main/crates/youki-wamr-executor

**Comment by 0xE282B0** [comment]
Ok, I remember where the problems were 😅.

## Rust Crate
There is a [wamr-sys](https://crates.io/crates/wamr-sys) crate by @rahul-thakoor on crates.io.
Unfortunately it is a bit outdated, but @jturcotte forked it and updated it. However, it did not have WASI enabled, so I forked it again and enabled WASI in [this repo](https://github.com/0xE282B0/wamr-sys).

The `wamr-sys` crate is just a bunch of rust-bindgen generated bindings. This leads to a lot of `unsafe` code. I put together a quick draft to outline the problems: https://github.com/0xE282B0/runwasi/pull/5

## Runwasi
The drafted shim contains a [test case which runs the wasi-demo-app](https://github.com/0xE282B0/runwasi/pull/5/files#diff-91800d108be73643bcffa3d3b8365792b84990e9ed1ad778aa32ccdadf1bad0eR112-R116):
```rust
#[test]
fn test_exec() -> anyhow::Result<()> {
    exec("../../target/wasm32-wasi/debug/wasi-demo-app.wasm");
    Ok(())
}
```
This already works, it also worked in the youki executor I tried some time ago. However, it does not work in the runwasi shim. because `wasm_runtime_init()` fails. (probably during the initialization of the memory allocator). We could try to play with the init configuration and may get someone from the WAMR project to get some insights on that problem.

## TLDR
In my opinion we would need to:
- [ ] (Re)publish the `wamr-sys` crate
- [ ] Create a `wamr` crate with a safe api
- [ ] Debug why the WAMR runtime initialization fails
- [ ] Create a PR to the runwasi project for the WAMR shim 🎉


I'm afraid that's a little more than I can spend working on open source projects. 🫣

Is there anyone with a need for WAMR containers, already?
Or is anyone interested in debugging WAMR on runwasi with libcontainer?

**Comment by Mossaka** [comment]
I want to reach out to maintainers from Wamr and asking if the community is interested in investing in this work.

**Comment by lum1n0us** [comment]
Wamr community is very eager to have volunteers to support it. Wamr team would like to provide all technical guidance beside Rust language knowledge.

**Comment by squillace** [comment]
@lum1n0us do you all have insights about https://github.com/containerd/runwasi/issues/337#issuecomment-1737317266? Is there another path or are some of those things you can help us with?

**Comment by Mossaka** [comment]
This one is completed with https://github.com/containerd/runwasi/pull/716
